File: CSC412/A3/tail.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <deque>

void print_last_10_lines(std::istream &input, const std::string &filename, bool print_filename) {
    std::string line;
    std::deque<std::string> last_10_lines;

    while (std::getline(input, line)) {
        if (last_10_lines.size() == 10) {
            last_10_lines.pop_front();
        }
        last_10_lines.push_back(line);
    }

    if (print_filename) {
        std::cout << "==> " << filename << " <==" << std::endl;
    }

    for (const std::string &line : last_10_lines) {
        std::cout << line << std::endl;
    }
}

int main(int argc, char *argv[]) {
    if (argc == 1) {
        print_last_10_lines(std::cin, "", false);
        return 0;
    }

    bool multiple_files = (argc > 2);

    for (int i = 1; i < argc; i++) {
        std::ifstream file(argv[i]);
        if (!file) {
            std::cerr << "tail: cannot open " << "'" << argv[i] << "'" << " for reading: No such file or directory" << std::endl;
            continue;
        }
        print_last_10_lines(file, argv[i], multiple_files);
        if (i < argc - 1) {
            std::cout << std::endl;
        }
    }
    return 0;
}

File: CSC412/assignments/A3/tail.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader};
use std::collections::VecDeque;

const MAX_LINES: usize = 10;

fn process_in_last10<R: BufRead>(reader: R, filename: Option<&str>, print_filename: bool) {
    let mut last_10_lines = VecDeque::new();
    for line in reader.lines() {
        match line {
            Ok(line) => {
                if last_10_lines.len() > MAX_LINES {
                    last_10_lines.pop_front();
                }
                last_10_lines.push_back(line);
            }
            Err(_) => {
                eprintln!("tail: error: could not read from file");
                return;
            }
        }
    }
    if print_filename {
        println!("==> {} <==", filename.unwrap());
    }  

    for line in last_10_lines {
        println!("{}", line);
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() == 1 {
        let stdin = io::stdin();
        let reader = stdin.lock();
        process_in_last10(reader, None, false);
    } else {
        let muli_files = args.len() > 2;
        let mut first = true;

        for filename in &args[1..] {
            match File::open(filename) {
                Ok(file) => {
                    let reader = BufReader::new(file);
                    if muli_files {
                        if !first {
                            println!();
                        }
                        first = false;
                    }
                    process_in_last10(reader, Some(filename), muli_files);
                }
                Err(_) => {
                    eprintln!("tail: error: cannot open '{}' for reading: No such file or directory", filename);
                }
            }
        }
    }
}